File: PRNumberBiz/app/src/main/java/com/pplus/prnumberbiz/apps/post/data/ReplyAdapter.java
package com.pplus.prnumberbiz.apps.post.data;

import android.content.Context;
import android.content.Intent;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageView;
import android.widget.TextView;

import com.bumptech.glide.Glide;
import com.bumptech.glide.request.RequestOptions;
import com.pple.pplus.utils.part.utils.time.DateFormatUtils;
import com.pplus.prnumberbiz.Const;
import com.pplus.prnumberbiz.R;
import com.pplus.prnumberbiz.apps.common.builder.AlertBuilder;
import com.pplus.prnumberbiz.apps.common.builder.OnAlertResultListener;
import com.pplus.prnumberbiz.apps.common.mgmt.LoginInfoManager;
import com.pplus.prnumberbiz.apps.common.ui.base.BaseArrayAdapter;
import com.pplus.prnumberbiz.apps.post.ui.PostDetailActivity;
import com.pplus.prnumberbiz.apps.post.ui.ReplyActivity;
import com.pplus.prnumberbiz.apps.post.ui.ReplyDetailActivity;
import com.pplus.prnumberbiz.apps.post.ui.ReplyEditActivity;
import com.pplus.prnumberbiz.core.code.common.EnumData;
import com.pplus.prnumberbiz.core.network.model.dto.Comment;
import com.pplus.prnumberbiz.core.util.PplusCommonUtil;

import java.text.SimpleDateFormat;
import java.util.ArrayList;
import java.util.Calendar;
import java.util.Date;
import java.util.Locale;
import java.util.Map;


/**
 * Created by j2n on 2016. 12. 22..
 */

public class ReplyAdapter extends BaseArrayAdapter<Comment>{

    private Map<Long, ArrayList<Comment>> cCommentList;
    private int mTodayYear, mTodayMonth, mTodayDay;
    private String mPostType;

    public ReplyAdapter(Context context){

        super(context, R.layout.item_reply);
        Calendar c = Calendar.getInstance();
        mTodayYear = c.get(Calendar.YEAR);
        mTodayMonth = c.get(Calendar.MONTH);
        mTodayDay = c.get(Calendar.DAY_OF_MONTH);
    }

    public void setPostType(String postType){

        this.mPostType = postType;
    }

    public void setcCommentList(Map<Long, ArrayList<Comment>> cCommentList){

        this.cCommentList = cCommentList;
    }

    @Override
    public View getView(int position, View convertView, ViewGroup parent){

        ViewHolder holder;

        if(convertView == null) {

            convertView = LayoutInflater.from(parent.getContext()).inflate(R.layout.item_reply, parent, false);
            holder = new ViewHolder();
            holder.initialize(convertView);
            convertView.setTag(holder);
        } else {
            holder = (ViewHolder) convertView.getTag();
        }

        final Comment item = getItem(position);



        if(item.isDeleted()) {
            holder.text_name.setText(R.string.word_unknown);
            holder.text_contents.setText(R.string.msg_delete_comment);
            holder.image_profileImage.setImageResource(R.drawable.img_post_profile_default);
        } else {
            if(item.getAuthor().getNo().equals(LoginInfoManager.getInstance().getUser().getNo())){
                holder.text_name.setText("" + LoginInfoManager.getInstance().getUser().getPage().getName());
                if(LoginInfoManager.getInstance().getUser().getPage().getProfileImage() != null) {
                    Glide.with(getContext()).load(LoginInfoManager.getInstance().getUser().getPage().getProfileImage().getUrl()).apply(new RequestOptions().centerCrop().placeholder(R.drawable.img_post_profile_default).error(R.drawable.img_post_profile_default)).into(holder.image_profileImage);
                }else{
                    holder.image_profileImage.setImageResource(R.drawable.img_post_profile_default);
                }

            }else{
                holder.text_name.setText("" + item.getAuthor().getNickname());
                if(item.getAuthor().getProfileImage() != null) {
                    Glide.with(getContext()).load(item.getAuthor().getProfileImage().getUrl()).apply(new RequestOptions().centerCrop().placeholder(R.drawable.img_post_profile_default).error(R.drawable.img_post_profile_default)).into(holder.image_profileImage);
                }else{
                    holder.image_profileImage.setImageResource(R.drawable.img_post_profile_default);
                }
            }

            holder.text_contents.setText("" + item.getComment());
        }


        if(cCommentList.get(item.getNo()) != null) {
            holder.layout_child.setVisibility(View.VISIBLE);

            int replyCount = cCommentList.get(item.getNo()).size();
            Comment childComment = cCommentList.get(item.getNo()).get(cCommentList.get(item.getNo()).size() - 1);
            if(childComment.getAuthor().getProfileImage() != null) {
                Glide.with(getContext()).load(childComment.getAuthor().getProfileImage().getUrl()).apply(new RequestOptions().centerCrop().placeholder(R.drawable.img_post_profile_default).error(R.drawable.img_post_profile_default)).into(holder.image_child_profileImage);
            }else{
                holder.image_child_profileImage.setImageResource(R.drawable.img_post_profile_default);
            }

            holder.text_child_name.setText("" + childComment.getAuthor().getNickname());
            holder.text_child_contents.setText("" + childComment.getComment());
            holder.text_child_replyCount.setText(getContext().getString(R.string.format_word_count_reply_of_reply, replyCount));

        } else {
            holder.layout_child.setVisibility(View.GONE);
        }

        try {
            Date d = DateFormatUtils.PPLUS_DATE_FORMAT.parse(item.getRegDate());
            Calendar c = Calendar.getInstance();
            c.setTime(d);

            int year = c.get(Calendar.YEAR);
            int month = c.get(Calendar.MONTH);
            int day = c.get(Calendar.DAY_OF_MONTH);

            if(mTodayYear == year && mTodayMonth == month && mTodayDay == day) {
                SimpleDateFormat output = new SimpleDateFormat("a HH:mm");
                holder.text_regDate.setText(output.format(d));
            } else {
                SimpleDateFormat output = new SimpleDateFormat("yyyy/MM/dd");
                holder.text_regDate.setText(output.format(d));
            }

        } catch (Exception e) {

        }

        holder.layout_child.setOnClickListener(new View.OnClickListener(){

            @Override
            public void onClick(View view){

                Intent intent = new Intent(getContext(), ReplyDetailActivity.class);
                intent.putExtra(Const.REPLY, item);

                if(cCommentList.get(item.getNo()) != null) {
                    intent.putParcelableArrayListExtra(Const.REPLY_CHILD, cCommentList.get(item.getNo()));
                }
                if(getContext() instanceof PostDetailActivity) {
                    ((PostDetailActivity) getContext()).startActivityForResult(intent, Const.REQ_REPLY);
                } else if(getContext() instanceof ReplyActivity) {
                    ((ReplyActivity) getContext()).startActivityForResult(intent, Const.REQ_REPLY);
                }

            }
        });

        if(item.getAuthor().getNo().equals(LoginInfoManager.getInstance().getUser().getNo())){
            holder.text_reply.setVisibility(View.GONE);
        }else{
            holder.text_reply.setVisibility(View.VISIBLE);
        }

        holder.text_reply.setOnClickListener(new View.OnClickListener(){

            @Override
            public void onClick(View view){


                Intent intent = new Intent(getContext(), ReplyDetailActivity.class);
                intent.putExtra(Const.REPLY, item);

                if(cCommentList.get(item.getNo()) != null) {
                    intent.putParcelableArrayListExtra(Const.REPLY_CHILD, cCommentList.get(item.getNo()));
                }
                if(getContext() instanceof PostDetailActivity) {
                    ((PostDetailActivity) getContext()).startActivityForResult(intent, Const.REQ_REPLY);
                } else if(getContext() instanceof ReplyActivity) {
                    ((ReplyActivity) getContext()).startActivityForResult(intent, Const.REQ_REPLY);
                }

            }
        });

        holder.layout_reply_parent.setOnClickListener(new View.OnClickListener(){

            @Override
            public void onClick(View view){

                final AlertBuilder.Builder builder = new AlertBuilder.Builder();
                builder.setLeftText(getContext().getString(R.string.word_cancel));

                final boolean isMe = LoginInfoManager.getInstance().getUser().getNo().equals(item.getAuthor().getNo());

                if(isMe) {
                    builder.setContents(getContext().getString(R.string.word_modified), getContext().getString(R.string.word_delete));
                } else {
                    builder.setContents(getContext().getString(R.string.msg_report));
                }

                builder.setStyle_alert(AlertBuilder.STYLE_ALERT.LIST_BOTTOM).setOnAlertResultListener(new OnAlertResultListener(){

                    @Override
                    public void onCancel(){

                    }

                    @Override
                    public void onResult(AlertBuilder.EVENT_ALERT event_alert){

                        switch (event_alert.getValue()) {
                            case 1:
                                if(isMe) {

                                    Intent intent = new Intent(getContext(), ReplyEditActivity.class);
                                    intent.putExtra(Const.REPLY, item);
                                    if(getContext() instanceof PostDetailActivity) {
                                        ((PostDetailActivity) getContext()).startActivityForResult(intent, Const.REQ_REPLY);
                                    }else if(getContext() instanceof ReplyActivity) {
                                        ((ReplyActivity) getContext()).startActivityForResult(intent, Const.REQ_REPLY);
                                    }

                                } else {
                                    PplusCommonUtil.Companion.report(getContext(), EnumData.REPORT_TYPE.comment, item.getNo());
                                }
                                break;
                            case 2:
                                if(isMe) {

                                    new AlertBuilder.Builder().setContents(getContext().getString(R.string.msg_question_delete_reply)).setLeftText(getContext().getString(R.string.word_cancel)).setRightText(getContext().getString(R.string.word_confirm)).setOnAlertResultListener(new OnAlertResultListener(){

                                        @Override
                                        public void onCancel(){

                                        }

                                        @Override
                                        public void onResult(AlertBuilder.EVENT_ALERT event_alert){

                                            switch (event_alert) {
                                                case RIGHT:
                                                    if(getContext() instanceof PostDetailActivity) {
                                                        ((PostDetailActivity) getContext()).deleteComment(item.getNo());
                                                    }else if(getContext() instanceof ReplyActivity) {
                                                        ((ReplyActivity) getContext()).deleteComment(item.getNo());
                                                    }

                                                    break;
                                            }
                                        }
                                    }).builder().show(getContext());

                                }
                                break;
                        }
                    }
                }).builder().show(getContext());
            }
        });


        return convertView;
    }

    class ViewHolder{

        public ImageView image_profileImage, image_child_profileImage;
        public TextView text_name, text_contents, text_regDate, text_child_name, text_child_contents, text_child_replyCount;
        private View text_reply, layout_reply_parent, layout_child;
        public Context context;

        public void initialize(View itemView){

            context = itemView.getContext();
            layout_reply_parent = itemView.findViewById(R.id.layout_reply_parent);
            image_profileImage = (ImageView) itemView.findViewById(R.id.image_reply_profileImage);
            image_child_profileImage = (ImageView) itemView.findViewById(R.id.image_reply_child_profileImage);
            text_name = (TextView) itemView.findViewById(R.id.text_reply_name);
            text_contents = (TextView) itemView.findViewById(R.id.text_reply_contents);
            text_regDate = (TextView) itemView.findViewById(R.id.text_reply_regDate);
            text_child_name = (TextView) itemView.findViewById(R.id.text_reply_child_name);
            text_child_contents = (TextView) itemView.findViewById(R.id.text_reply_child_contents);
            text_child_replyCount = (TextView) itemView.findViewById(R.id.text_reply_child_replyCount);
            text_reply = itemView.findViewById(R.id.text_reply_reply);
            layout_child = itemView.findViewById(R.id.layout_reply_child);
            layout_child.setVisibility(View.GONE);
        }
    }
}

File: PRNumber/app/src/main/java/com/pplus/prnumberuser/apps/post/data/ReplyAdapter.kt
//package com.pplus.prnumberuser.apps.post.data
//
//import android.app.Activity
//import android.content.Context
//import android.content.Intent
//import androidx.recyclerview.widget.RecyclerView
//import android.view.LayoutInflater
//import android.view.View
//import android.view.ViewGroup
//import androidx.fragment.app.FragmentActivity
//import com.bumptech.glide.Glide
//import com.bumptech.glide.request.RequestOptions
//import com.pplus.utils.part.utils.time.DateFormatUtils
//import com.pplus.prnumberuser.Const
//import com.pplus.prnumberuser.R
//import com.pplus.prnumberuser.apps.common.builder.AlertBuilder
//import com.pplus.prnumberuser.apps.common.builder.OnAlertResultListener
//import com.pplus.prnumberuser.apps.common.mgmt.LoginInfoManager
//import com.pplus.prnumberuser.apps.common.ui.base.BaseActivity
//import com.pplus.prnumberuser.apps.post.ui.PostReplyActivity
//import com.pplus.prnumberuser.apps.post.ui.ReplyDetailActivity
//import com.pplus.prnumberuser.apps.post.ui.ReplyEditActivity
//import com.pplus.prnumberuser.apps.post.ui.ReviewWriteActivity
//import com.pplus.prnumberuser.core.code.common.EnumData
//import com.pplus.prnumberuser.core.network.ApiBuilder
//import com.pplus.prnumberuser.core.network.model.dto.Comment
//import com.pplus.prnumberuser.core.network.model.dto.Post
//import com.pplus.prnumberuser.core.network.model.response.NewResultResponse
//import com.pplus.prnumberuser.core.util.PplusCommonUtil
//import com.pplus.prnumberuser.core.util.ToastUtil
//import kotlinx.android.synthetic.main.item_reply.view.*
//import com.pplus.networks.common.PplusCallback
//import retrofit2.Call
//import java.text.SimpleDateFormat
//import java.util.*
//
///**
// * Created by imac on 2018. 1. 8..
// */
//class ReplyAdapter : RecyclerView.Adapter<ReplyAdapter.ViewHolder> {
//
//    var mContext: Context? = null
//    var mDataList: MutableList<Comment>? = null
//    var listener: OnItemClickListener? = null
//    var cCommentList: Map<Long, ArrayList<Comment>>? = null
//    val mTodayYear: Int = 0
//    val mTodayMonth: Int = 0
//    val mTodayDay: Int = 0
//    var mPost: Post? = null
//
//    interface OnItemClickListener {
//
//        fun onItemClick(position: Int)
//    }
//
//    constructor(context: Context, post: Post) : super() {
//        this.mContext = context
//        this.mDataList = ArrayList()
//        this.mPost = post
//    }
//
//    fun setcCommentList(cCommentList: Map<Long, ArrayList<Comment>>) {
//
//        this.cCommentList = cCommentList
//    }
//
//    fun setCommentCount(count: Int) {
//        if (mPost != null) {
//            mPost!!.commentCount = count
//        }
//
//    }
//
//    fun setOnItemClickListener(listener: OnItemClickListener) {
//
//        this.listener = listener
//    }
//
//    fun getItem(position: Int): Comment {
//
//        return mDataList!!.get(position)
//    }
//
//    fun getDataList(): MutableList<Comment>? {
//
//        return mDataList
//    }
//
//    fun add(data: Comment) {
//
//        if (mDataList == null) {
//            mDataList = ArrayList<Comment>()
//        }
//        mDataList!!.add(data)
//        notifyDataSetChanged()
//    }
//
//    fun addAll(dataList: List<Comment>) {
//
//        if (this.mDataList == null) {
//            this.mDataList = ArrayList<Comment>()
//        }
//
//        this.mDataList!!.addAll(dataList)
//        notifyDataSetChanged()
//    }
//
//    fun replaceData(position: Int, data: Comment) {
//
//        mDataList!!.removeAt(position)
//        mDataList!!.add(position, data)
//        notifyDataSetChanged()
//    }
//
//    fun clear() {
//
//        mDataList = ArrayList<Comment>()
//        notifyDataSetChanged()
//    }
//
//    fun setDataList(dataList: MutableList<Comment>) {
//
//        this.mDataList = dataList
//        notifyDataSetChanged()
//    }
//
//    class ViewHolder(itemView: View) : androidx.recyclerview.widget.RecyclerView.ViewHolder(itemView) {
//
//        val layout_reply_parent = itemView.layout_reply_parent
//        val image_profileImage = itemView.image_reply_profileImage
//        val image_child_profileImage = itemView.image_reply_child_profileImage
//        var text_name = itemView.text_reply_name
//        val text_contents = itemView.text_reply_contents
//        var text_regDate = itemView.text_reply_regDate
//        val text_child_name = itemView.text_reply_child_name
//        val text_child_contents = itemView.text_reply_child_contents
//        val text_child_replyCount = itemView.text_reply_child_replyCount
//        val text_reply = itemView.text_reply_reply
//        val layout_child = itemView.layout_reply_child
//
//        init {
//
//        }
//    }
//
//    override fun getItemCount(): Int {
//        return mDataList!!.size
//    }
//
//    override fun onBindViewHolder(holder: ViewHolder, position: Int) {
//        val item = mDataList!!.get(position);
//
//        if (item.isDeleted) {
//            holder.text_name.setText(R.string.word_unknown)
//            holder.text_contents.setText(R.string.msg_delete_comment)
//            holder.image_profileImage.setImageResource(R.drawable.img_post_profile_default)
//        } else {
//
//            if (item.author.useStatus == EnumData.UseStatus.leave.name) {
//                holder.text_name.setText(R.string.word_unknown)
//                holder.image_profileImage.setImageResource(R.drawable.img_post_profile_default)
//            } else {
//                holder.text_name.text = item.author.nickname
//                if (item.author.profileImage != null) {
//                    Glide.with(mContext!!).load(item.author.profileImage!!.url).apply(RequestOptions().centerCrop().placeholder(R.drawable.img_post_profile_default).error(R.drawable.img_post_profile_default)).into(holder.image_profileImage)
//                } else {
//                    holder.image_profileImage.setImageResource(R.drawable.img_post_profile_default)
//                }
//            }
//
//            holder.text_contents.text = item.comment
//        }
//
//        if (cCommentList != null && cCommentList!![item.no] != null) {
//            holder.layout_child.visibility = View.VISIBLE
//
//            val replyCount = cCommentList!![item.no]!!.size
//            val childComment = cCommentList!![item.no]!![cCommentList!![item.no]!!.size - 1]
//
//            if (childComment.author.useStatus == EnumData.UseStatus.leave.name) {
//                holder.text_child_name.setText(R.string.word_unknown)
//            } else {
//                holder.text_child_name.text = childComment.author.nickname
//
//                if (childComment.author.profileImage != null) {
//                    Glide.with(mContext!!).load(childComment.author.profileImage!!.url).apply(RequestOptions().centerCrop().placeholder(R.drawable.img_post_profile_default).error(R.drawable.img_post_profile_default)).into(holder.image_child_profileImage)
//                } else {
//                    holder.image_child_profileImage.setImageResource(R.drawable.img_post_profile_default)
//                }
//            }
//
//            holder.text_child_name.text = childComment.author.nickname
//
//            holder.text_child_contents.text = childComment.comment
//            holder.text_child_replyCount.text = mContext?.getString(R.string.format_word_count_reply_of_reply, replyCount)
//
//        } else {
//            holder.layout_child.visibility = View.GONE
//        }
//
//        try {
//            val d = DateFormatUtils.PPLUS_DATE_FORMAT.parse(item.regDate)
//            val c = Calendar.getInstance()
//            c.time = d
//
//            val year = c.get(Calendar.YEAR)
//            val month = c.get(Calendar.MONTH)
//            val day = c.get(Calendar.DAY_OF_MONTH)
//
//            if (mTodayYear == year && mTodayMonth == month && mTodayDay == day) {
//                val output = SimpleDateFormat("a HH:mm")
//                holder.text_regDate.text = output.format(d)
//            } else {
//                val output = SimpleDateFormat("yyyy/MM/dd")
//                holder.text_regDate.text = output.format(d)
//            }
//
//        } catch (e: Exception) {
//            holder.text_regDate.text = ""
//        }
//
//
//        holder.layout_child.setOnClickListener {
//            val intent = Intent(mContext, ReplyDetailActivity::class.java)
//            intent.putExtra(Const.POST, mPost)
//            intent.putExtra(Const.REPLY, item)
//
//            if (cCommentList!![item.no] != null) {
//                intent.putParcelableArrayListExtra(Const.REPLY_CHILD, cCommentList!![item.no])
//            }
//            (holder.itemView.context as PostReplyActivity).replyLauncher.launch(intent)
//        }
//
//        if (LoginInfoManager.getInstance().isMember && item.author.no == LoginInfoManager.getInstance().user.no) {
//            holder.text_reply.visibility = View.GONE
//        } else {
//            holder.text_reply.visibility = View.VISIBLE
//        }
//
//        holder.text_reply.setOnClickListener {
//            val intent = Intent(mContext, ReplyDetailActivity::class.java)
//            intent.putExtra(Const.POST, mPost)
//            intent.putExtra(Const.REPLY, item)
//
//            if (cCommentList!![item.no] != null) {
//                intent.putParcelableArrayListExtra(Const.REPLY_CHILD, cCommentList!![item.no])
//            }
//            (holder.itemView.context as PostReplyActivity).replyLauncher.launch(intent)
//        }
//
//        holder.layout_reply_parent.setOnClickListener {
//
//            if (!PplusCommonUtil.loginCheck((holder.itemView.context as FragmentActivity), null)) {
//                return@setOnClickListener
//            }
//
//            val builder = AlertBuilder.Builder()
//            builder.setLeftText(mContext?.getString(R.string.word_cancel))
//
//            val isMe = LoginInfoManager.getInstance().user.no == item.author.no
//
//            if (isMe) {
//                builder.setContents(mContext?.getString(R.string.word_modified), mContext?.getString(R.string.word_delete))
//            } else {
//                builder.setContents(mContext?.getString(R.string.msg_report))
//            }
//
//            builder.setStyle_alert(AlertBuilder.STYLE_ALERT.LIST_BOTTOM).setOnAlertResultListener(object : OnAlertResultListener {
//
//                override fun onCancel() {
//
//                }
//
//                override fun onResult(event_alert: AlertBuilder.EVENT_ALERT) {
//
//                    when (event_alert.getValue()) {
//                        1 -> if (isMe) {
//
//                            val intent = Intent(mContext, ReplyEditActivity::class.java)
//                            intent.putExtra(Const.REPLY, item)
//                            (holder.itemView.context as PostReplyActivity).replyLauncher.launch(intent)
//
//                        } else {
//                            PplusCommonUtil.report(mContext!!, EnumData.REPORT_TYPE.comment, item.no)
//                        }
//                        2 -> if (isMe) {
//
//                            AlertBuilder.Builder().setContents(mContext?.getString(R.string.msg_question_delete_reply)).setLeftText(mContext?.getString(R.string.word_cancel)).setRightText(mContext?.getString(R.string.word_confirm)).setOnAlertResultListener(object : OnAlertResultListener {
//
//                                override fun onCancel() {
//
//                                }
//
//                                override fun onResult(event_alert: AlertBuilder.EVENT_ALERT) {
//
//                                    when (event_alert) {
//                                        AlertBuilder.EVENT_ALERT.RIGHT -> if (mContext is PostReplyActivity) {
//                                            (mContext as PostReplyActivity).deleteComment(item.no)
//                                        }
//                                    }
//                                }
//                            }).builder().show(mContext)
//
//                        }
//                    }
//                }
//            }).builder().show(mContext)
//        }
//
//    }
//
//    private fun deletePost(no: Long, position: Int) {
//
//        ApiBuilder.create().deletePost(no).setCallback(object : PplusCallback<NewResultResponse<Any>> {
//
//            override fun onResponse(call: Call<NewResultResponse<Any>>, response: NewResultResponse<Any>) {
//                mDataList?.removeAt(position)
//                notifyItemRemoved(position)
//
//            }
//
//            override fun onFailure(call: Call<NewResultResponse<Any>>, t: Throwable, response: NewResultResponse<Any>) {
//                ToastUtil.showAlert(mContext, R.string.msg_can_not_delete_review)
//
//            }
//        }).build().call()
//    }
//
//    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): ViewHolder {
//        val v = LayoutInflater.from(parent.context).inflate(R.layout.item_reply, parent, false)
//        return ViewHolder(v)
//    }
//}